File: ios/Nebenkosten/Kern/Heizkosten.swift
import Foundation

enum Brennstoff: String, Codable, CaseIterable, Identifiable {
    case erdgas
    case heizoel
    case fluessiggas
    case pellets
    case fernwaerme
    case waermepumpe

    var id: String { rawValue }

    /// Heizwert in kWh je Mengeneinheit (Richtwert für Überschlagsrechnungen).
    var heizwert: Double {
        switch self {
        case .erdgas: return 10.0
        case .heizoel: return 10.0
        case .fluessiggas: return 6.57
        case .pellets: return 4.8
        case .fernwaerme: return 1.0
        case .waermepumpe: return 1.0
        }
    }

    var mengeneinheit: String {
        switch self {
        case .erdgas: return "m³"
        case .heizoel, .fluessiggas: return "l"
        case .pellets: return "kg"
        case .fernwaerme, .waermepumpe: return "kWh"
        }
    }

    var bezeichnung: String {
        switch self {
        case .erdgas: return "Erdgas H"
        case .heizoel: return "Heizöl EL"
        case .fluessiggas: return "Flüssiggas"
        case .pellets: return "Holzpellets"
        case .fernwaerme: return "Fernwärme"
        case .waermepumpe: return "Wärmepumpenstrom"
        }
    }

    var beschriftung: String {
        "\(bezeichnung) (\(Geld.zahl(heizwert, 2)) kWh/\(mengeneinheit))"
    }
}

enum Warmwassermethode: String, Codable, CaseIterable, Identifiable {
    case wmz
    case formel
    case prozent

    var id: String { rawValue }
    var bezeichnung: String {
        switch self {
        case .wmz: return "gemessene Wärmemenge (Wärmemengenzähler)"
        case .formel: return "rechnerisch nach § 9 Abs. 2 HeizkostenV"
        case .prozent: return "fester Prozentsatz"
        }
    }
}

/// Abrechnung der Heiz- und Warmwasserkosten nach der Heizkostenverordnung.
///
/// - § 7 Abs. 1: 50 bis 70 % der Heizkosten nach erfasstem Verbrauch, der Rest
///   als Grundkosten nach Wohnfläche.
/// - § 8 Abs. 1: dasselbe für die Warmwasserkosten.
/// - § 9 Abs. 2: Bei verbundenen Anlagen wird die auf das Warmwasser
///   entfallende Wärmemenge vorrangig gemessen, ersatzweise gilt
///   `Q = 2,5 · V · (tw − 10) / 1000` in MWh.
/// - § 12 Abs. 1: Ohne verbrauchsabhängige Abrechnung darf der Nutzer seinen
///   Anteil um 15 % kürzen.
enum Heizkosten {

    static let verbrauchsanteilMin = 0.5
    static let verbrauchsanteilMax = 0.7

    static func verbrauchsanteilZulaessig(_ anteil: Double) -> Bool {
        anteil.isFinite && anteil >= verbrauchsanteilMin - 1e-9 && anteil <= verbrauchsanteilMax + 1e-9
    }

    /// Wärmemenge des Warmwassers in kWh nach § 9 Abs. 2 HeizkostenV.
    static func warmwasserWaermemengeKwh(volumenM3: Double, temperaturC: Double = 60) -> Double {
        guard volumenM3 > 0, temperaturC > 10 else { return 0 }
        // Q [MWh] = 2,5 · V · (tw − 10) / 1000  ⇒  Q [kWh] = 2,5 · V · (tw − 10)
        return 2.5 * volumenM3 * (temperaturC - 10)
    }

    // MARK: - Warmwasseranteil

    struct Warmwasseranteil {
        var anteil: Double = 0
        var methode: String = ""
        var warmwasserKwh: Double = 0
        var gesamtwaermeKwh: Double = 0
    }

    struct Warmwassereingaben {
        var methode: Warmwassermethode = .formel
        var warmwasserKwh: Double = 0
        var volumen: Double = 0
        var temperatur: Double = 60
        var prozentsatz: Double = 0.18
        var gesamtwaermeKwh: Double = 0
        var brennstoffmenge: Double = 0
        var brennstoff: Brennstoff = .erdgas
    }

    static func warmwasseranteil(_ eingaben: Warmwassereingaben) -> Warmwasseranteil {
        if eingaben.methode == .prozent {
            return Warmwasseranteil(
                anteil: klemme(eingaben.prozentsatz, 0, 1),
                methode: "Fester vertraglich oder technisch begründeter Anteil")
        }

        let gesamtwaerme = eingaben.gesamtwaermeKwh > 0
            ? eingaben.gesamtwaermeKwh
            : eingaben.brennstoffmenge * eingaben.brennstoff.heizwert

        let warmwasser = eingaben.methode == .wmz
            ? eingaben.warmwasserKwh
            : warmwasserWaermemengeKwh(volumenM3: eingaben.volumen, temperaturC: eingaben.temperatur)

        guard gesamtwaerme > 0, warmwasser > 0 else {
            return Warmwasseranteil(
                anteil: 0,
                methode: "Keine Aufteilung möglich – Datengrundlage fehlt",
                warmwasserKwh: warmwasser,
                gesamtwaermeKwh: gesamtwaerme)
        }

        return Warmwasseranteil(
            anteil: klemme(warmwasser / gesamtwaerme, 0, 1),
            methode: eingaben.methode == .wmz
                ? "Gemessene Wärmemenge Warmwasser (Wärmemengenzähler, § 9 Abs. 2 Satz 1 HeizkostenV)"
                : "Rechnerisch nach § 9 Abs. 2 HeizkostenV: Q = 2,5 · V · (tw − 10) / 1000",
            warmwasserKwh: warmwasser,
            gesamtwaermeKwh: gesamtwaerme)
    }

    private static func klemme(_ wert: Double, _ unten: Double, _ oben: Double) -> Double {
        guard wert.isFinite else { return unten }
        return Swift.min(oben, Swift.max(unten, wert))
    }

    // MARK: - Verteilung

    struct Nutzer {
        let id: String
        let bezeichnung: String
        var leerstand: Bool = false
        var flaecheTage: Double = 0
        var verbrauchHeizung: Double = 0
        var verbrauchWarmwasser: Double = 0
    }

    struct Zeile: Identifiable {
        let id: String
        let bezeichnung: String
        let leerstand: Bool
        let flaecheTage: Double
        let flaecheAnteil: Double
        let verbrauchHeizung: Double
        let verbrauchHeizungAnteil: Double
        let verbrauchWarmwasser: Double
        let verbrauchWarmwasserAnteil: Double
        let heizGrund: Cent
        let heizVerbrauch: Cent
        let wwGrund: Cent
        let wwVerbrauch: Cent
        let kuerzung15: Cent
        let summe: Cent
    }

    struct Toepfe {
        var heizGrund: Cent = 0
        var heizVerbrauch: Cent = 0
        var wwGrund: Cent = 0
        var wwVerbrauch: Cent = 0
    }

    struct Bezugsgroessen {
        var flaecheTageGesamt: Double = 0
        var verbrauchHeizungGesamt: Double = 0
        var verbrauchWarmwasserGesamt: Double = 0
    }

    struct Ergebnis {
        var brennstoffBrutto: Cent = 0
        var nebenkosten: Cent = 0
        var kostenGesamtRoh: Cent = 0
        var co2 = CO2.Aufteilung()
        var umlagefaehigeBrennstoffkosten: Cent = 0
        var gesamtUmlagefaehig: Cent = 0
        var verbunden = false
        var warmwasserAufteilung = Warmwasseranteil()
        var kostenHeizung: Cent = 0
        var kostenWarmwasser: Cent = 0
        var anteilVerbrauchHeizung: Double = 0.7
        var anteilVerbrauchWarmwasser: Double = 0.7
        var verbrauchserfassung = true
        var toepfe = Toepfe()
        var bezug = Bezugsgroessen()
        var zeilen: [Zeile] = []
        var summeVerteilt: Cent = 0
    }

    struct Eingaben {
        var kosten = Heizkostenpositionen()
        var co2Kosten: Cent = 0
        var co2EmissionKg: Double = 0
        var gebaeudetyp: Gebaeudetyp = .wohn
        var co2Ausnahme = false
        var verbunden = true
        var kostenWarmwasserSeparat: Cent = 0
        var warmwasser = Warmwassereingaben()
        var anteilVerbrauchHeizung: Double = 0.7
        var anteilVerbrauchWarmwasser: Double = 0.7
        var verbrauchserfassung = true
        var tageZeitraum: Int = 365
        var wohnflaecheGesamt: Double = 0
        var nutzer: [Nutzer] = []
    }

    static func berechne(_ eingaben: Eingaben) -> Ergebnis {
        var ergebnis = Ergebnis()

        let brennstoff = eingaben.kosten.brennstoff
        let nebenkosten = eingaben.kosten.nebenkosten

        ergebnis.brennstoffBrutto = brennstoff
        ergebnis.nebenkosten = nebenkosten
        ergebnis.kostenGesamtRoh = brennstoff + nebenkosten

        // 1. CO₂-Kosten aufteilen – der Vermieteranteil wird vorab abgezogen.
        ergebnis.co2 = CO2.teile(
            kostenGesamt: eingaben.co2Kosten,
            emissionKg: eingaben.co2EmissionKg,
            wohnflaeche: eingaben.wohnflaecheGesamt,
            tageZeitraum: eingaben.tageZeitraum,
            gebaeudetyp: eingaben.gebaeudetyp,
            ausnahme: eingaben.co2Ausnahme)

        ergebnis.umlagefaehigeBrennstoffkosten = max(0, brennstoff - ergebnis.co2.vermieterCent)
        ergebnis.gesamtUmlagefaehig = ergebnis.umlagefaehigeBrennstoffkosten + nebenkosten

        // 2. Aufteilung Heizung / Warmwasser
        ergebnis.verbunden = eingaben.verbunden
        ergebnis.warmwasserAufteilung = eingaben.verbunden
            ? warmwasseranteil(eingaben.warmwasser)
            : Warmwasseranteil(anteil: 0, methode: "Getrennte Anlagen – keine Aufteilung erforderlich")

        ergebnis.kostenWarmwasser = eingaben.verbunden
            ? Geld.anteil(von: ergebnis.gesamtUmlagefaehig, ergebnis.warmwasserAufteilung.anteil)
            : eingaben.kostenWarmwasserSeparat
        ergebnis.kostenHeizung = eingaben.verbunden
            ? ergebnis.gesamtUmlagefaehig - ergebnis.kostenWarmwasser
            : ergebnis.gesamtUmlagefaehig

        // 3. Grund- und Verbrauchsanteile bilden
        let aH = klemme(eingaben.anteilVerbrauchHeizung, 0, 1)
        let aW = klemme(eingaben.anteilVerbrauchWarmwasser, 0, 1)
        ergebnis.anteilVerbrauchHeizung = aH
        ergebnis.anteilVerbrauchWarmwasser = aW
        ergebnis.verbrauchserfassung = eingaben.verbrauchserfassung

        var toepfe = Toepfe()
        toepfe.heizVerbrauch = Geld.anteil(von: ergebnis.kostenHeizung, aH)
        toepfe.heizGrund = ergebnis.kostenHeizung - toepfe.heizVerbrauch
        toepfe.wwVerbrauch = Geld.anteil(von: ergebnis.kostenWarmwasser, aW)
        toepfe.wwGrund = ergebnis.kostenWarmwasser - toepfe.wwVerbrauch
        ergebnis.toepfe = toepfe

        // 4. Bezugsgrößen
        var bezug = Bezugsgroessen()
        bezug.flaecheTageGesamt = eingaben.nutzer.map(\.flaecheTage).summe
        bezug.verbrauchHeizungGesamt = eingaben.nutzer.map(\.verbrauchHeizung).summe
        bezug.verbrauchWarmwasserGesamt = eingaben.nutzer.map(\.verbrauchWarmwasser).summe
        ergebnis.bezug = bezug

        let ohneErfassung = !eingaben.verbrauchserfassung

        ergebnis.zeilen = eingaben.nutzer.map { nutzer in
            let flAnteil = bezug.flaecheTageGesamt > 0 ? nutzer.flaecheTage / bezug.flaecheTageGesamt : 0
            let vhAnteil = bezug.verbrauchHeizungGesamt > 0 ? nutzer.verbrauchHeizung / bezug.verbrauchHeizungGesamt : 0
            let vwAnteil = bezug.verbrauchWarmwasserGesamt > 0 ? nutzer.verbrauchWarmwasser / bezug.verbrauchWarmwasserGesamt : 0

            // Ohne Verbrauchserfassung fällt alles auf den Flächenmaßstab zurück.
            let heizVerbrauchAnteil = ohneErfassung ? flAnteil : vhAnteil
            let wwVerbrauchAnteil = ohneErfassung ? flAnteil : vwAnteil

            let heizGrund = Geld.anteil(von: toepfe.heizGrund, flAnteil)
            let heizVerbrauch = Geld.anteil(von: toepfe.heizVerbrauch, heizVerbrauchAnteil)
            let wwGrund = Geld.anteil(von: toepfe.wwGrund, flAnteil)
            let wwVerbrauch = Geld.anteil(von: toepfe.wwVerbrauch, wwVerbrauchAnteil)

            let roh = heizGrund + heizVerbrauch + wwGrund + wwVerbrauch
            // § 12 Abs. 1 HeizkostenV: 15 % Kürzung ohne verbrauchsabhängige Abrechnung.
            let kuerzung = (ohneErfassung && !nutzer.leerstand) ? Geld.runde(Double(roh) * 0.15) : 0

            return Zeile(
                id: nutzer.id,
                bezeichnung: nutzer.bezeichnung,
                leerstand: nutzer.leerstand,
                flaecheTage: nutzer.flaecheTage,
                flaecheAnteil: flAnteil,
                verbrauchHeizung: nutzer.verbrauchHeizung,
                verbrauchHeizungAnteil: heizVerbrauchAnteil,
                verbrauchWarmwasser: nutzer.verbrauchWarmwasser,
                verbrauchWarmwasserAnteil: wwVerbrauchAnteil,
                heizGrund: heizGrund,
                heizVerbrauch: heizVerbrauch,
                wwGrund: wwGrund,
                wwVerbrauch: wwVerbrauch,
                kuerzung15: kuerzung,
                summe: roh - kuerzung)
        }

        ergebnis.summeVerteilt = ergebnis.zeilen.map(\.summe).summe
        return ergebnis
    }
}

File: src/core/heizkosten.js
/**
 * Abrechnung der Heiz- und Warmwasserkosten nach der Verordnung über
 * Heizkostenabrechnung (HeizkostenV).
 *
 * Kernregeln:
 *  - § 7 Abs. 1: 50 bis 70 % der Heizkosten sind nach erfasstem Verbrauch zu
 *    verteilen, der Rest als Grundkosten nach Wohn-/Nutzfläche oder umbautem Raum.
 *  - § 8 Abs. 1: dasselbe für die Warmwasserkosten.
 *  - § 9 Abs. 2: Bei verbundenen Anlagen wird die auf das Warmwasser
 *    entfallende Wärmemenge vorrangig gemessen; ersatzweise gilt
 *    Q = 2,5 · V · (tw − 10) / 1000  [MWh].
 *  - § 12 Abs. 1: Wird nicht verbrauchsabhängig abgerechnet, kann der Nutzer
 *    den auf ihn entfallenden Anteil um 15 % kürzen.
 */

import { runde, anteilVon, summe } from './money.js';
import { teileCo2Kosten, EMISSIONSFAKTOR } from './co2.js';

/**
 * Energieträger einer Wärmeerzeugung.
 *
 *  - `heizwert`      kWh je Mengeneinheit (Standardwert für Überschlagsrechnungen)
 *  - `co2Pflichtig`  true, wenn der Träger dem BEHG unterliegt und damit
 *                    CO2-Kosten nach dem CO2KostAufG anfallen. Strom fällt
 *                    nicht darunter (er unterliegt dem europäischen
 *                    Emissionshandel), Holz ebenfalls nicht.
 *  - `arbeitszahl`   true, wenn aus der eingesetzten Energie erst über eine
 *                    Arbeitszahl die erzeugte Wärmemenge wird (Wärmepumpe).
 */
export const ENERGIETRAEGER = {
  erdgas: { bezeichnung: 'Erdgas H', einheit: 'm³', heizwert: 10.0, co2Pflichtig: true },
  heizoel: { bezeichnung: 'Heizöl EL', einheit: 'l', heizwert: 10.0, co2Pflichtig: true },
  fluessiggas: { bezeichnung: 'Flüssiggas', einheit: 'l', heizwert: 6.57, co2Pflichtig: true },
  pellets: { bezeichnung: 'Holzpellets', einheit: 'kg', heizwert: 4.8, co2Pflichtig: false },
  fernwaerme: { bezeichnung: 'Fernwärme', einheit: 'kWh', heizwert: 1.0, co2Pflichtig: true },
  waermepumpe: { bezeichnung: 'Wärmepumpenstrom', einheit: 'kWh', heizwert: 1.0, co2Pflichtig: false, arbeitszahl: true },
  heizstrom: { bezeichnung: 'Heizstrom (direkt)', einheit: 'kWh', heizwert: 1.0, co2Pflichtig: false },
};

export function energietraeger(id) {
  return ENERGIETRAEGER[id] || null;
}

/** Emissionsfaktor in kg CO₂ je kWh Endenergie. */
export function emissionsfaktor(traegerId) {
  return EMISSIONSFAKTOR[traegerId] || 0;
}

export const VERBRAUCHSANTEIL_MIN = 0.5;
export const VERBRAUCHSANTEIL_MAX = 0.7;

/**
 * Prüft den gewählten Verbrauchsanteil gegen § 7 Abs. 1 / § 8 Abs. 1 HeizkostenV.
 */
export function verbrauchsanteilZulaessig(anteil) {
  return Number.isFinite(anteil) && anteil >= VERBRAUCHSANTEIL_MIN - 1e-9 && anteil <= VERBRAUCHSANTEIL_MAX + 1e-9;
}

/**
 * Wärmemenge des Warmwassers nach § 9 Abs. 2 HeizkostenV.
 * @param {number} volumenM3 verbrauchtes Warmwasser in m³
 * @param {number} temperaturC mittlere Warmwassertemperatur in °C
 * @returns {number} Wärmemenge in kWh
 */
export function warmwasserWaermemengeKwh(volumenM3, temperaturC = 60) {
  const v = Number.isFinite(volumenM3) ? volumenM3 : 0;
  const t = Number.isFinite(temperaturC) ? temperaturC : 60;
  if (v <= 0 || t <= 10) return 0;
  // Q [MWh] = 2,5 · V · (tw − 10) / 1000  =>  Q [kWh] = 2,5 · V · (tw − 10)
  return 2.5 * v * (t - 10);
}

/* ---------------------------------------------------------- Wärmeerzeuger */

/**
 * Wärmemenge eines Erzeugers in kWh.
 *
 * Vorrang hat die Messung: die Differenz zweier Zählerstände, ersatzweise eine
 * direkt eingetragene Wärmemenge. Erst wenn kein Wärmemengenzähler vorhanden
 * ist, wird aus der eingesetzten Energiemenge gerechnet – bei einer Wärmepumpe
 * über die Arbeitszahl, sonst über den Heizwert.
 */
export function waermemengeVon(e = {}) {
  const stand = (e.zaehler?.standEnde || 0) - (e.zaehler?.standAnfang || 0);
  if (stand > 0) return stand;
  if (e.waermemengeKwh > 0) return e.waermemengeKwh;

  const t = ENERGIETRAEGER[e.energietraeger];
  const endenergie = (e.menge || 0) * (t?.heizwert || 0);
  if (t?.arbeitszahl) return endenergie * (e.arbeitszahl > 0 ? e.arbeitszahl : 1);
  return endenergie;
}

/** true, wenn die Wärmemenge dieses Erzeugers gemessen und nicht gerechnet ist. */
export function istGemessen(e = {}) {
  return (e.zaehler?.standEnde || 0) - (e.zaehler?.standAnfang || 0) > 0 || (e.waermemengeKwh || 0) > 0;
}

/**
 * Bereitet die Wärmeerzeuger für die Abrechnung auf und bildet die Summen.
 *
 * CO2-Kosten werden nur von Erzeugern übernommen, deren Energieträger dem
 * BEHG unterliegt. Wärmepumpenstrom bleibt dabei außen vor – für ihn gilt das
 * CO2KostAufG nicht, und er senkt dadurch zugleich den Emissionskennwert des
 * Gebäudes.
 */
export function fasseErzeugerZusammen(liste = []) {
  const erzeuger = liste.map((e) => {
    const t = ENERGIETRAEGER[e.energietraeger] || {};
    const co2Pflichtig = !!t.co2Pflichtig;
    return {
      id: e.id,
      bezeichnung: e.bezeichnung || t.bezeichnung || 'Wärmeerzeuger',
      energietraeger: e.energietraeger,
      traegerBezeichnung: t.bezeichnung || e.energietraeger || '',
      mengeneinheit: t.einheit || '',
      menge: e.menge || 0,
      arbeitszahl: t.arbeitszahl ? (e.arbeitszahl > 0 ? e.arbeitszahl : 1) : 0,
      zaehlernummer: e.zaehler?.nummer || '',
      waermeKwh: waermemengeVon(e),
      gemessen: istGemessen(e),
      kostenCent: e.kostenCent || 0,
      co2Pflichtig,
      co2KostenCent: co2Pflichtig ? e.co2KostenCent || 0 : 0,
      co2EmissionKg: co2Pflichtig ? e.co2EmissionKg || 0 : 0,
    };
  });

  const waermeGesamtKwh = summe(erzeuger.map((e) => e.waermeKwh));
  return {
    erzeuger,
    anzahl: erzeuger.length,
    hybrid: erzeuger.length > 1,
    kostenGesamt: summe(erzeuger.map((e) => e.kostenCent)),
    waermeGesamtKwh,
    vollstaendigGemessen: erzeuger.length > 0 && erzeuger.every((e) => e.gemessen),
    co2KostenGesamt: summe(erzeuger.map((e) => e.co2KostenCent)),
    co2EmissionGesamt: summe(erzeuger.map((e) => e.co2EmissionKg)),
    anteile: erzeuger.map((e) => ({
      id: e.id,
      bezeichnung: e.bezeichnung,
      anteilWaerme: waermeGesamtKwh > 0 ? e.waermeKwh / waermeGesamtKwh : 0,
    })),
  };
}

/**
 * Ermittelt den Anteil der Warmwasserbereitung an den Brennstoffkosten
 * einer verbundenen Anlage.
 *
 * @param {object} p
 * @param {'wmz'|'formel'|'prozent'} p.modus
 * @param {number} p.warmwasserKwh      gemessene Wärmemenge Warmwasser (Modus 'wmz')
 * @param {number} p.warmwasserVolumen  Gesamtwarmwasser in m³ (Modus 'formel')
 * @param {number} p.temperatur         mittlere Warmwassertemperatur in °C
 * @param {number} p.gesamtwaermeKwh    insgesamt erzeugte Wärmemenge in kWh
 * @param {number} p.prozentsatz        fester Anteil 0..1 (Modus 'prozent')
 */
export function warmwasserAnteil(p = {}) {
  const modus = p.modus || 'formel';

  if (modus === 'prozent') {
    const anteil = klemme(p.prozentsatz ?? 0.18, 0, 1);
    return { anteil, methode: 'Fester vertraglich/technisch begründeter Anteil', warmwasserKwh: 0, gesamtwaermeKwh: 0 };
  }

  const gesamtwaermeKwh = p.gesamtwaermeKwh > 0 ? p.gesamtwaermeKwh : 0;

  const warmwasserKwh =
    modus === 'wmz'
      ? p.warmwasserKwh || 0
      : warmwasserWaermemengeKwh(p.warmwasserVolumen, p.temperatur);

  if (gesamtwaermeKwh <= 0 || warmwasserKwh <= 0) {
    return { anteil: 0, methode: 'Keine Aufteilung möglich – Datengrundlage fehlt', warmwasserKwh, gesamtwaermeKwh };
  }

  return {
    anteil: klemme(warmwasserKwh / gesamtwaermeKwh, 0, 1),
    methode:
      modus === 'wmz'
        ? 'Gemessene Wärmemenge Warmwasser (Wärmemengenzähler, § 9 Abs. 2 Satz 1 HeizkostenV)'
        : 'Rechnerisch nach § 9 Abs. 2 HeizkostenV: Q = 2,5 · V · (tw − 10) / 1000',
    warmwasserKwh,
    gesamtwaermeKwh,
  };
}

function klemme(wert, min, max) {
  if (!Number.isFinite(wert)) return min;
  return Math.min(max, Math.max(min, wert));
}

/**
 * Vollständige Heiz- und Warmwasserkostenabrechnung.
 *
 * @param {object} p
 * @param {Array}  p.erzeuger          Wärmeerzeuger: [{ energietraeger, menge, arbeitszahl, zaehler, waermemengeKwh, kostenCent, co2KostenCent, co2EmissionKg }]
 * @param {object} p.kosten            Cent-Beträge: betriebsstrom, wartung, messdienst, schornsteinfeger, sonstiges
 * @param {object} p.co2               Gebäudeangaben für das CO2KostAufG: gebaeudetyp, ausnahme
 * @param {boolean} p.verbunden        verbundene Anlage (Heizung + Warmwasser aus einer Erzeugung)
 * @param {object} p.warmwasser        Parameter für warmwasserAnteil()
 * @param {number} p.anteilVerbrauchHeizung   0,5..0,7
 * @param {number} p.anteilVerbrauchWarmwasser 0,5..0,7
 * @param {boolean} p.verbrauchserfassung  false => § 12 HeizkostenV, 15 % Kürzung
 * @param {Array} p.nutzer  [{ id, bezeichnung, flaecheTage, verbrauchHeizung, verbrauchWarmwasser, leerstand }]
 * @param {number} p.tageZeitraum
 * @param {number} p.wohnflaecheGesamt
 */
export function berechneHeizkosten(p) {
  const kosten = p.kosten || {};
  const nutzer = p.nutzer || [];

  // 0. Wärmeerzeuger zusammenfassen. Eine Hybridanlage – etwa Gaskessel und
  //    Wärmepumpe – liefert hier mehrere Einträge; ihre Kosten bilden gemeinsam
  //    die Brennstoffkosten, ihre Wärmemengen die Gesamtwärmemenge.
  const erzeugung = fasseErzeugerZusammen(p.erzeuger || []);

  const brennstoffBrutto = erzeugung.kostenGesamt;
  const nebenkosten = summe([
    kosten.betriebsstrom,
    kosten.wartung,
    kosten.messdienst,
    kosten.schornsteinfeger,
    kosten.sonstiges,
  ]);

  // 1. CO2-Kosten aufteilen — der Vermieteranteil wird vorab abgezogen
  //    (§ 6 CO2KostAufG). Berücksichtigt werden nur Energieträger, die dem
  //    BEHG unterliegen; Wärmepumpenstrom bleibt außen vor.
  const co2 = teileCo2Kosten({
    co2KostenCent: erzeugung.co2KostenGesamt,
    emissionKg: erzeugung.co2EmissionGesamt,
    wohnflaeche: p.wohnflaecheGesamt || 0,
    tageZeitraum: p.tageZeitraum || 365,
    gebaeudetyp: p.co2?.gebaeudetyp || 'wohn',
    ausnahme: !!p.co2?.ausnahme,
  });

  const umlagefaehigeBrennstoffkosten = Math.max(0, brennstoffBrutto - co2.vermieterCent);
  const gesamtUmlagefaehig = umlagefaehigeBrennstoffkosten + nebenkosten;

  // 2. Aufteilung Heizung / Warmwasser
  const ww = p.verbunden
    ? warmwasserAnteil({ ...(p.warmwasser || {}), gesamtwaermeKwh: erzeugung.waermeGesamtKwh })
    : { anteil: 0, methode: 'Getrennte Anlagen – keine Aufteilung erforderlich', warmwasserKwh: 0, gesamtwaermeKwh: 0 };

  const kostenWarmwasser = p.verbunden ? anteilVon(gesamtUmlagefaehig, ww.anteil) : (p.kostenWarmwasserSeparat || 0);
  const kostenHeizung = p.verbunden ? gesamtUmlagefaehig - kostenWarmwasser : gesamtUmlagefaehig;

  // 3. Grund- und Verbrauchsanteile bilden (§ 7 Abs. 1, § 8 Abs. 1 HeizkostenV)
  const aH = klemme(p.anteilVerbrauchHeizung ?? 0.7, 0, 1);
  const aW = klemme(p.anteilVerbrauchWarmwasser ?? 0.7, 0, 1);

  const heizVerbrauchTopf = anteilVon(kostenHeizung, aH);
  const heizGrundTopf = kostenHeizung - heizVerbrauchTopf;
  const wwVerbrauchTopf = anteilVon(kostenWarmwasser, aW);
  const wwGrundTopf = kostenWarmwasser - wwVerbrauchTopf;

  // 4. Bezugsgrößen
  const flaecheTageGesamt = summe(nutzer.map((n) => n.flaecheTage || 0));
  const verbrauchHeizungGesamt = summe(nutzer.map((n) => n.verbrauchHeizung || 0));
  const verbrauchWarmwasserGesamt = summe(nutzer.map((n) => n.verbrauchWarmwasser || 0));

  const zeilen = nutzer.map((n) => {
    const flAnteil = flaecheTageGesamt > 0 ? (n.flaecheTage || 0) / flaecheTageGesamt : 0;
    const vhAnteil = verbrauchHeizungGesamt > 0 ? (n.verbrauchHeizung || 0) / verbrauchHeizungGesamt : 0;
    const vwAnteil = verbrauchWarmwasserGesamt > 0 ? (n.verbrauchWarmwasser || 0) / verbrauchWarmwasserGesamt : 0;

    // Ohne Verbrauchserfassung fällt alles auf den Flächenmaßstab zurück.
    const ohneErfassung = p.verbrauchserfassung === false;
    const heizGrund = anteilVon(heizGrundTopf, flAnteil);
    const heizVerbrauch = anteilVon(heizVerbrauchTopf, ohneErfassung ? flAnteil : vhAnteil);
    const wwGrund = anteilVon(wwGrundTopf, flAnteil);
    const wwVerbrauch = anteilVon(wwVerbrauchTopf, ohneErfassung ? flAnteil : vwAnteil);

    const roh = heizGrund + heizVerbrauch + wwGrund + wwVerbrauch;
    // § 12 Abs. 1 HeizkostenV: 15 % Kürzung, wenn nicht verbrauchsabhängig abgerechnet wird.
    const kuerzung = ohneErfassung && !n.leerstand ? runde(roh * 0.15) : 0;

    return {
      id: n.id,
      bezeichnung: n.bezeichnung,
      leerstand: !!n.leerstand,
      flaecheTage: n.flaecheTage || 0,
      flaecheAnteil: flAnteil,
      verbrauchHeizung: n.verbrauchHeizung || 0,
      verbrauchHeizungAnteil: ohneErfassung ? flAnteil : vhAnteil,
      verbrauchWarmwasser: n.verbrauchWarmwasser || 0,
      verbrauchWarmwasserAnteil: ohneErfassung ? flAnteil : vwAnteil,
      heizGrund,
      heizVerbrauch,
      wwGrund,
      wwVerbrauch,
      kuerzung15: kuerzung,
      summe: roh - kuerzung,
    };
  });

  return {
    kostenRoh: { brennstoffBrutto, nebenkosten, gesamt: brennstoffBrutto + nebenkosten },
    erzeugung,
    co2,
    umlagefaehigeBrennstoffkosten,
    gesamtUmlagefaehig,
    verbunden: !!p.verbunden,
    warmwasserAufteilung: ww,
    kostenHeizung,
    kostenWarmwasser,
    anteilVerbrauchHeizung: aH,
    anteilVerbrauchWarmwasser: aW,
    verbrauchserfassung: p.verbrauchserfassung !== false,
    toepfe: { heizGrundTopf, heizVerbrauchTopf, wwGrundTopf, wwVerbrauchTopf },
    bezug: { flaecheTageGesamt, verbrauchHeizungGesamt, verbrauchWarmwasserGesamt },
    zeilen,
    summeVerteilt: summe(zeilen.map((z) => z.summe)),
  };
}
